invalidate SVG <use> element when inline style changes

nullhook opened the pull request:
#### 0823259833f9ebda926b59b2cb59563c17fa064e
<pre>
invalidate SVG &lt;use&gt; element when inline style changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=201918">https://bugs.webkit.org/show_bug.cgi?id=201918</a>
<a href="https://rdar.apple.com/98577657">rdar://98577657</a>

Reviewed by Antti Koivisto.

SVG &lt;use&gt; elements create shadow trees that clone referenced elements,
but dynamic style changes made through element.style.property did
not invalidate these shadow trees. This PR properly invalidates it.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/StyledElement.cpp:
(WebCore::StyledElement::invalidateStyleAttribute):

Canonical link: <a href="https://commits.webkit.org/299112@main">https://commits.webkit.org/299112@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/954aa6f155da0e793940852b9f2ab877870837ac

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117921 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37599 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28232 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124066 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69956 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f0838ae0-d7b6-4fad-a592-3b0fc0c6eadf) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119799 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38292 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46181 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89486 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59099 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/35350cd2-df7d-4c80-8421-410f49f10238) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120873 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30494 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105737 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69982 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f7eb22ef-85bf-4bd3-9eb4-aeb7a0f59290) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29560 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23853 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67730 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99915 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24032 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127148 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44824 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33764 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98156 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45185 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101962 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97944 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24910 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43338 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21317 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41257 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44696 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50370 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44156 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47501 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45845 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/dom/StyledElement.cpp:
```diff
@@ -192,6 +193,11 @@ void StyledElement::invalidateStyleAttribute()
+        if (auto* svgElement = dynamicDowncast<SVGElement>(this))
+            SVGElement::InstanceInvalidationGuard guard(*svgElement);
+    }
+
```
You can just call `invalidateInstances()` directly instead of creating a pointless InstanceInvalidationGuard

nullhook replied:
`::invalidateInstances()` is private. however, i can do something like below which is exactly the same body of invalidateInstances but it looks repetitive. 

```
if (auto* svgElement = dynamicDowncast<SVGElement>(this)) {
      if (svgElement->instanceUpdatesBlocked())
          return;
  
      for (auto& instance : copyToVectorOf<Ref<SVGElement>>(svgElement->instances())) {
          if (RefPtr useElement = instance->correspondingUseElement())
              useElement->invalidateShadowTree();
          instance->setCorrespondingElement(nullptr);
      }
  }
...
```

nullhook replied:
made `invalidateInstances` public

nullhook committed:
invalidate SVG <use> element when inline style changes
https://bugs.webkit.org/show_bug.cgi?id=201918
rdar://98577657

Reviewed by Antti Koivisto.

SVG <use> elements create shadow trees that clone referenced elements,
but dynamic style changes made through element.style.property did
not invalidate these shadow trees. This PR properly invalidates it.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/StyledElement.cpp:
(WebCore::StyledElement::invalidateStyleAttribute):

Canonical link: https://commits.webkit.org/299112@main

shallawa commented on Source/WebCore/dom/StyledElement.cpp:
```diff
@@ -192,6 +193,11 @@ void StyledElement::invalidateStyleAttribute()
 
     Node::invalidateStyle(validity);
 
+    if (isSVGElement()) {
+        if (auto* svgElement = dynamicDowncast<SVGElement>(this))
+            svgElement->invalidateInstances();
+    }
```
Should not `StyledElement ::invalidateStyleAttribute()` be made virtual? In this case this code can be moved to `SVGElement::invalidateStyleAttribute()` after calling the base `StyledElement ::invalidateStyleAttribute()`?
